src/olmo_core/nn/attention/__init__.py

```diff
@@ -929,3 +1004,100 @@ def repeat_kv(x: torch.Tensor, n_rep: int) -> torch.Tensor:
         .expand(bs, slen, n_kv_heads, n_rep, head_dim)
         .reshape(bs, slen, n_kv_heads * n_rep, head_dim)
     )
+
+
+def _get_flex_attn_mask_mod(
+    window_size: Optional[Tuple[int, int]] = None,
+    doc_lens: Optional[Tuple[int, ...]] = None,
+    device: Optional[torch.device] = None,
+) -> Callable[[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor], torch.Tensor]:
+    mask_mods = []
+
+    def _causal_mask_mod(
+        B: torch.Tensor, H: torch.Tensor, q_idx: torch.Tensor, kv_idx: torch.Tensor
+    ) -> torch.Tensor:
+        return q_idx >= kv_idx
+
+    mask_mods.append(_causal_mask_mod)
+
+    if window_size is not None and window_size != (-1, -1):
+
+        def _sliding_window_mask_mod(
```

Comment:
The key value prop of FlexAttn is that its easy to experiment with a variety of different attention types. I think we should try to expose a config interface that allows users to define the set of masks and the score mod they want to use.

Reply:
This is actually coming up currently with @aman-17 's work trying to add in GPT-OSS style attention sinks